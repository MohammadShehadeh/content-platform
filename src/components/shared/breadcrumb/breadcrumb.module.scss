.breadcrumb {
  display: flex;
  align-items: center;
  gap: var(--space-xs);
  font-size: var(--font-size-sm);
  color: hsl(var(--muted-foreground));
}

.breadcrumbList {
  display: flex;
  align-items: center;
  list-style: none;
  margin: 0;
  padding: 0;
}

.item {
  display: inline-flex;
  align-items: center;
  gap: var(--space-xs);

  &:hover {
    color: hsl(var(--foreground));
  }

  &.current {
    color: hsl(var(--foreground));
    font-weight: var(--font-weight-medium);
    pointer-events: none;
  }
}

.listItem {
  display: inline-flex;
  align-items: center;
}

.separator {
  display: flex;
  align-items: center;
  color: hsl(var(--muted-foreground));
  margin: 0 var(--space-xs);
}
