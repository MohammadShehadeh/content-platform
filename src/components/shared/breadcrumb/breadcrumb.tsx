'use client';

import { Slot } from '@radix-ui/react-slot';
import type { PropsWithChildren } from 'react';
import React from 'react';

import { cn } from '@/lib/utils';

import styles from './breadcrumb.module.scss';

interface BreadcrumbContextValue {
  separator: React.ReactNode;
}

const BreadcrumbContext = React.createContext<BreadcrumbContextValue>({
  separator: '/',
});

const useBreadcrumb = () => {
  const context = React.useContext(BreadcrumbContext);
  if (!context) {
    throw new Error('useBreadcrumb must be used within a Breadcrumb component');
  }
  return context;
};

interface BreadcrumbProps extends PropsWithChildren {
  separator?: React.ReactNode;
  className?: string;
}

export const Breadcrumb = ({ children, separator = '/', className, ...props }: BreadcrumbProps) => {
  return (
    <BreadcrumbContext.Provider value={{ separator }}>
      <nav className={cn(styles.breadcrumb, className)} aria-label="Breadcrumb" {...props}>
        <ol className={styles.breadcrumbList}>{children}</ol>
      </nav>
    </BreadcrumbContext.Provider>
  );
};

interface BreadcrumbItemProps extends PropsWithChildren {
  href?: string;
  current?: boolean;
  className?: string;
}

export const BreadcrumbItem = ({ children, current, className, ...props }: BreadcrumbItemProps) => {
  const { separator } = useBreadcrumb();

  return (
    <li className={styles.listItem}>
      <span
        className={cn(
          styles.item,
          {
            [styles.current]: current,
          },
          className
        )}
        aria-current={current ? 'page' : undefined}
        {...props}
      >
        {children}
      </span>
      {!current && <span className={styles.separator}>{separator}</span>}
    </li>
  );
};

interface BreadcrumbLinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement>, PropsWithChildren {
  asChild?: boolean;
  current?: boolean;
}

export const BreadcrumbLink = ({ children, asChild, current, className, ...props }: BreadcrumbLinkProps) => {
  const { separator } = useBreadcrumb();
  const Comp = asChild ? Slot : 'a';

  return (
    <li className={styles.listItem}>
      <Comp
        className={cn(
          styles.item,
          {
            [styles.current]: current,
          },
          className
        )}
        aria-current={current ? 'page' : undefined}
        {...props}
      >
        {children}
      </Comp>
      {!current && <span className={styles.separator}>{separator}</span>}
    </li>
  );
};

interface BreadcrumbSeparatorProps {
  children?: React.ReactNode;
  className?: string;
}

export const BreadcrumbSeparator = ({ children, className, ...props }: BreadcrumbSeparatorProps) => {
  const { separator } = React.useContext(BreadcrumbContext);

  return (
    <span className={cn(styles.separator, className)} {...props}>
      {children || separator}
    </span>
  );
};
