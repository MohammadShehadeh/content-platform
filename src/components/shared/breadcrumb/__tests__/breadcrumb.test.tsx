import { render, screen } from '@testing-library/react';

import { Breadcrumb, BreadcrumbItem, BreadcrumbLink } from '@/components/shared/breadcrumb';

describe('Breadcrumb', () => {
  it('should render breadcrumb items correctly', () => {
    render(
      <Breadcrumb>
        <BreadcrumbLink href="/">Home</BreadcrumbLink>
        <BreadcrumbLink href="/posts">Posts</BreadcrumbLink>
        <BreadcrumbItem current>Post Details</BreadcrumbItem>
      </Breadcrumb>
    );

    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByText('Posts')).toBeInTheDocument();
    expect(screen.getByText('Post Details')).toBeInTheDocument();
  });

  it('should render with custom separator', () => {
    render(
      <Breadcrumb separator=">">
        <BreadcrumbLink href="/">Home</BreadcrumbLink>
        <BreadcrumbItem current>Posts</BreadcrumbItem>
      </Breadcrumb>
    );

    expect(screen.getByText('>')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <Breadcrumb>
        <BreadcrumbLink href="/">Home</BreadcrumbLink>
        <BreadcrumbItem current>Posts</BreadcrumbItem>
      </Breadcrumb>
    );

    expect(container).toMatchSnapshot();
  });
});
