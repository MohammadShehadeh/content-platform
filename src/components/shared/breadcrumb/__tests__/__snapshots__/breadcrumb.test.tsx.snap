// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb matches snapshot 1`] = `
<div>
  <nav
    aria-label="Breadcrumb"
    class="breadcrumb"
  >
    <ol
      class="breadcrumbList"
    >
      <li
        class="listItem"
      >
        <a
          class="item"
          href="/"
        >
          Home
        </a>
        <span
          class="separator"
        >
          /
        </span>
      </li>
      <li
        class="listItem"
      >
        <span
          aria-current="page"
          class="item current"
        >
          Posts
        </span>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb should match snapshot 1`] = `
<div>
  <nav
    aria-label="Breadcrumb"
    class="breadcrumb"
  >
    <ol
      class="breadcrumbList"
    >
      <li
        class="listItem"
      >
        <a
          class="item"
          href="/"
        >
          Home
        </a>
        <span
          class="separator"
        >
          /
        </span>
      </li>
      <li
        class="listItem"
      >
        <span
          aria-current="page"
          class="item current"
        >
          Posts
        </span>
      </li>
    </ol>
  </nav>
</div>
`;
